### PR TITLE
WIP for Review: Make atomic option enum

### PR DESCRIPTION
**Pull request by shahab-vahedi**

### CODE REVIEW

File: gcc/config/arc64/arc64-c.def [hunk 1]
```diff
@@ -17,7 +17,10 @@
    <http://www.gnu.org/licenses/>.
 */
 
-//ARC_C_DEF ("__ARC_UNALIGNED__", unaligned_access)
+ARC64_C_DEF ("__ARC64_ATOMIC_0__",	(arc_atomic_option == 0))
```
Comment (claziss) [inline gcc/config/arc64/arc64-c.def:20]: This macro makes no sense

File: gcc/config/arc64/arc64-c.def [hunk 2]
```diff
@@ -17,7 +17,10 @@
...
+ARC64_C_DEF ("__ARC64_ATOMIC_1__",	(arc_atomic_option == 1))
```
Comment (claziss) [inline gcc/config/arc64/arc64-c.def:21]: I would use a macro with value here (i.e. `__ARC64_ATOMIC__`). You can make this using:
`builtin_define_with_int_value ("__ARC64_ATOMIC__", arc64_atomic_option);` 
in `arc64-c.c` file.

File: gcc/config/arc64/arc64.h [hunk 3]
```diff
@@ -506,6 +506,17 @@ extern const enum reg_class arc64_regno_to_regclass[];
        "jl\t[r12]\n"						\
        TEXT_SECTION_ASM_OP);
 
+/* ATOMIC options.  */
+/* FIXME: is 0 okay or should it be -1 like DEFAULT_arc_mpy_option?  */
+/* Default atomic option value.  */
+#undef DEFAULT_ARC_ATOMIC_OPTION
+#define DEFAULT_ARC_ATOMIC_OPTION 0
+
+#define TARGET_ATOMIC_1 (arc_atomic_option > 0)
```
Comment (claziss) [inline gcc/config/arc64/arc64.h:515]: Please use `ARC64_HAS_ATOMIC_1` macro name. Likewise for all atomic options fallowing.

File: gcc/config/arc64/arc64.h [hunk 4]
```diff
@@ -506,6 +506,17 @@ extern const enum reg_class arc64_regno_to_regclass[];
...
+#define TARGET_ATOMIC_2 (arc_atomic_option > 1)
+#define TARGET_ATOMIC_3 (arc_atomic_option > 2)
+#define TARGET_ATOMIC   (TARGET_ATOMIC_1 | TARGET_ATOMIC_2 | TARGET_ATOMIC_3)
```
Comment (claziss) [inline gcc/config/arc64/arc64.h:518]: Not sure u need this, and by all means this is equivalent with `ARC64_HAS_ATOMIC_1` macro.

File: gcc/config/arc64/atomic.md [hunk 5]
```diff
@@ -58,7 +58,7 @@
    (match_operand:SI 5 "const_int_operand")		;; is_weak
    (match_operand:SI 6 "const_int_operand")		;; mod_s
    (match_operand:SI 7 "const_int_operand")]		;; mod_f
-  ""
+  "TARGET_ATOMIC"
```
Comment (claziss) [inline gcc/config/arc64/atomic.md:61]: This is valid for ATOMIC_1 upwards

File: gcc/config/arc64/atomic.md [hunk 6]
```diff
@@ -77,7 +77,7 @@
 	   (match_operand:SI 5 "const_int_operand")	;; mod_s
 	   (match_operand:SI 6 "const_int_operand")]	;; mod_f
 	  ARC64_VUNSPEC_CAS))]
-  ""
+  "TARGET_ATOMIC"
```
Comment (claziss) [inline gcc/config/arc64/atomic.md:80]: likewise.

File: gcc/config/arc64/atomic.md [hunk 7]
```diff
@@ -91,7 +91,7 @@
 	(unspec_volatile:SDI
 	  [(match_operand:SDI 1 "mem_noofs_operand" "ATOMC")]
 	  ARC64_VUNSPEC_LL))]
-  ""
+  "TARGET_ATOMIC"
```
Comment (claziss) [inline gcc/config/arc64/atomic.md:94]: This has to be be conditional. The 64 bit llockd is only valid from `ATOMIC_2+`.

File: gcc/config/arc64/atomic.md [hunk 8]
```diff
@@ -103,7 +103,7 @@
 	(unspec_volatile:SDI[(match_operand:SDI 1 "register_operand" "r")]
 			   ARC64_VUNSPEC_SC))
    (clobber (reg:CC_Z CC_REGNUM))]
-  ""
+  "TARGET_ATOMIC"
```
Comment (claziss) [inline gcc/config/arc64/atomic.md:106]: likewise

File: gcc/config/arc64/arc64.h [hunk 9]
```diff
@@ -506,6 +506,17 @@ extern const enum reg_class arc64_regno_to_regclass[];
...
        TEXT_SECTION_ASM_OP);
 
+/* ATOMIC options.  */
+/* FIXME: is 0 okay or should it be -1 like DEFAULT_arc_mpy_option?  */
+/* Default atomic option value.  */
+#undef DEFAULT_ARC_ATOMIC_OPTION
+#define DEFAULT_ARC_ATOMIC_OPTION 0
+
+#define TARGET_ATOMIC_1 (arc_atomic_option > 0)
+#define TARGET_ATOMIC_2 (arc_atomic_option > 1)
```
Comment (claziss) [inline gcc/config/arc64/arc64.h:516]: Moreover, we should use `arc64_atomic_option`.

File: gcc/config/arc64/atomic.md [hunk 10]
```diff
@@ -131,7 +131,7 @@
 			    ARC64_VUNSPEC_EX))
    (set (match_dup 1)
 	(match_operand:SDI 2 "register_operand" "0"))]
-  ""
+  "TARGET_ATOMIC"
```
Comment (claziss) [inline gcc/config/arc64/atomic.md:134]: AFAIK `ex` instructions are baseline, you need to remove the predicate here.

File: gcc/config/arc64/atomic.md [hunk 11]
```diff
@@ -192,7 +192,7 @@
    (atomicop:SDI (match_dup 1)
 		(match_operand:SDI 2 "register_operand" "")) ;; operand
    (match_operand:SDI 3 "const_int_operand" "")]	;; model
-  ""
+  "TARGET_ATOMIC"
```
Comment (claziss) [inline gcc/config/arc64/atomic.md:195]: The predicate needs to reflect the ATOMIC_2 option for the llockd/scondd variants

File: gcc/config/arc64/atomic.md [hunk 12]
```diff
@@ -204,7 +204,7 @@
    (match_operand:SDI 1 "mem_noofs_operand" "")		;; memory
    (match_operand:SDI 2 "register_operand" "")		;; operand
    (match_operand:SDI 3 "const_int_operand" "")]	;; model
-  ""
+  "TARGET_ATOMIC"
```
Comment (claziss) [inline gcc/config/arc64/atomic.md:207]: likewise.

**Review by claziss — CHANGES_REQUESTED** [review]

**Comment by shahab-vahedi** [comment]
Closed. See:

https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/pull/89
https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/pull/90

**Pull request closed without merging**